nav2_costmap_2d/src/costmap_subscriber.cpp

```diff
@@ -16,41 +16,82 @@
 #include <memory>
 
 #include "nav2_costmap_2d/costmap_subscriber.hpp"
+#include "nav2_costmap_2d/cost_values.hpp"
+
+using nav2_costmap_2d::NO_INFORMATION;
+using nav2_costmap_2d::LETHAL_OBSTACLE;
+using nav2_costmap_2d::FREE_SPACE;
 
 namespace nav2_costmap_2d
 {
 
 CostmapSubscriber::CostmapSubscriber(
   nav2_util::LifecycleNode::SharedPtr node,
-  const std::string & topic_name)
+  const std::string & topic_name,
+  bool use_raw)
 : CostmapSubscriber(node->get_node_base_interface(),
     node->get_node_topics_interface(),
     node->get_node_logging_interface(),
-    topic_name)
+    node->get_node_parameters_interface(),
+    topic_name,
+    use_raw)
 {}
 
 CostmapSubscriber::CostmapSubscriber(
   rclcpp::Node::SharedPtr node,
-  const std::string & topic_name)
+  const std::string & topic_name,
+  bool use_raw)
 : CostmapSubscriber(node->get_node_base_interface(),
     node->get_node_topics_interface(),
     node->get_node_logging_interface(),
-    topic_name)
+    node->get_node_parameters_interface(),
+    topic_name,
+    use_raw)
 {}
 
 CostmapSubscriber::CostmapSubscriber(
   const rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base,
   const rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr node_topics,
   const rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr node_logging,
-  const std::string & topic_name)
+  const rclcpp::node_interfaces::NodeParametersInterface::SharedPtr node_parameters,
+  const std::string & topic_name,
+  bool use_raw)
 : node_base_(node_base),
   node_topics_(node_topics),
   node_logging_(node_logging),
-  topic_name_(topic_name)
+  node_parameters_(node_parameters),
+  topic_name_(topic_name),
+  use_raw_(use_raw)
 {
-  costmap_sub_ = rclcpp::create_subscription<nav2_msgs::msg::Costmap>(node_topics_, topic_name_,
+  if (use_raw_) {
+    costmap_raw_sub_ = rclcpp::create_subscription<nav2_msgs::msg::Costmap>(node_topics_,
+        topic_name_,
+        rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+        std::bind(&CostmapSubscriber::costmapRawCallback, this, std::placeholders::_1));
+  } else {
+    costmap_sub_ = rclcpp::create_subscription<nav_msgs::msg::OccupancyGrid>(node_topics_,
+        topic_name_,
+        rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+        std::bind(&CostmapSubscriber::costmapCallback, this, std::placeholders::_1));
+    costmap_update_sub_ = rclcpp::create_subscription<map_msgs::msg::OccupancyGridUpdate>(
+      node_topics_,
+      topic_name_,
       rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
-      std::bind(&CostmapSubscriber::costmapCallback, this, std::placeholders::_1));
+      std::bind(&CostmapSubscriber::costmapUpdateCallback, this, std::placeholders::_1));
+    rclcpp::Parameter p;
+    if (node_parameters_->get_parameter("track_unknown_space", p)) {
```

Comment:
I don't see where the recoveries declare these parameters. 

Not being able to pull parameters from other nodes is a real pain in the ass.

Reply:
The recoveries don't declare them currently, though they could. Parameters can be pulled from other nodes if you create a parameter client object.

Reply:
You're  getting a parameter that hasn't been declared -- how does that work?

Reply:
This version of `get_parameter` will return a bool `False` if the parameter is not declared or not set. If it has not been declared, then the member variables will remain at their default value.

Reply:
That seems not ideal, since they'll essentially never be set.

Reply:
Do you have a different proposal? We could declare them if not declared in this code, which would minimally at least expose these parameters on the node if they didn't exist, but functionally it would be the same. This code is modified from the static layer and how it converts map data into the costmap layer, so someone could decide to match behavior with the current static layer or have a different value interpretation specifically for recoveries (e.g. choose a different lethal object threshold).

Reply:
Well the static layer is in the same node as its costmap so that should be fine, its getting the actual parameter as set by the user. 

I'm not sure what the best answer is, but don't you see where this is problematic? Its really breaking the encapsulation of the costmaps and now any process using this costmap subscriber object is going to need to define these parameters to be consistent with the costmap its subscribing to. Change a param in one place you need to change in `N`